Direct Assign IDXGISwapChain1 to the ISwapChainPanelNative in Babylon React Native(reworked)

nkari82 opened the pull request:
This pull request solves several problems on the Windows version.

1. Problems not being able to redraw when changing JavaScript.
2. Crash occurs when changing the window size in debug mode.
3. By creating a render-only thread, the React native UI is smooth.
4. Destructor not being called in EngineView.

ReactNative has a pre-pull request.
https://github.com/BabylonJS/BabylonNative/pull/606

chrisfromwork commented on Modules/@babylonjs/react-native/shared/BabylonNative.cpp:
```diff
@@ -123,7 +124,8 @@ namespace Babylon
 
-            m_graphics->RenderCurrentFrame();
+            if(m_graphics)
+				m_graphics->RenderCurrentFrame();
```
>				m_g [](start = 0, length = 7)

nit: we should probably do a pass and make sure all tabs are replaced with spaces

chrisfromwork replied:
the selected text for this comment looks a bit strange. The white space leading this m_graphics->RenderCurrentFrame() uses tabs compared to spaces, which is the actual thing we will want to update

chrisfromwork commented on Modules/@babylonjs/react-native-windows/windows/BabylonReactNative/EngineView.cpp:
```diff
@@ -2,173 +2,452 @@
+
+		// Clear the back buffer and depth stencil view.
+		float clearColor[4]{ 0.392156899f, 0.584313750f, 0.929411829f, 1.000000000f };
+		s_d3dContext->ClearRenderTargetView(_backBufferPtr.Get(), clearColor);
```
is this the blue background that can be seen when initializing? Would we prefer to have this as black?

chrisfromwork commented on Modules/@babylonjs/react-native-windows/windows/BabylonReactNative/EngineView.cpp:
```diff
@@ -2,173 +2,452 @@
+
+			// Store values so they can be accessed from a background thread.
+			_compositionScaleX = sender.CompositionScaleX();
+			_compositionScaleY = sender.CompositionScaleY();
```
When testing this locally, have you ever seen a composition scale != 1? When testing this out on my end I see the correct visual scaling when this composition scale has a value of 1. After initialization this value gets set to 2, which then causes the content to display incorrectly.

chrisfromwork replied:
For what its worth, the swap chain panel size changed event already reports width/height in pixels.

chrisfromwork replied:
it seems like there may be issues here related to: https://github.com/BabylonJS/BabylonNative/pull/582

When i compare this to existing behaviors, this composition scaling change now occurs after babylon.js is initialized. Previously, the composition scale was non one before initialization, which seemed to successfully get relayed to the engine. I'll continue to see whether we can figure out how to fix this.

chrisfromwork replied:
it seems like we may need to call Graphics::SetHardwareScalingLevel to submit this value

chrisfromwork commented:
![Screenshot (76)](https://user-images.githubusercontent.com/10225640/108751190-c45ae480-74f6-11eb-86fd-7c885ebab246.png)
I am seeing this sort of rendering behavior when composition scale == 2. I am not yet sure if we need to do something differently to recreate the device resources or if we need to avoid applying the composition scale all together.

chrisfromwork commented:
regardless, this is an awesome change. It seems like now toggling the engine view does not result in any crashes. :)

chrisfromwork commented on Modules/@babylonjs/react-native-windows/windows/BabylonReactNative/EngineView.cpp:
```diff
@@ -2,173 +2,452 @@
+			//	critical_section::scoped_lock lock(s_criticalSection);
+			//	Babylon::EnableView();
+			//}
+
```
should we file a follow up issue to look into reenabling this commented out code?

chrisfromwork commented on Modules/@babylonjs/react-native-windows/windows/BabylonReactNative/EngineView.cpp:
```diff
@@ -2,173 +2,452 @@
+		CreateSizeDependentResources();
+
+		auto swapChainPanel = static_cast<SwapChainPanel>(*this);
+		auto swapChainPanelNative = swapChainPanel.as<ISwapChainPanelNative>().get();;
```
>; [](start = 79, length = 1)

nit: remove extra ';'

chrisfromwork commented:
@rgerd @tymlipari @ryantrem @syntheticmagus @bghgary @Drigax can you take a look at this review?

ryantrem commented on Modules/@babylonjs/react-native/shared/BabylonNative.cpp:
```diff
@@ -142,7 +144,66 @@ namespace Babylon
+            }
+            else
+            {
+                JSGuard guard(m_jsDispatcher);
+                m_graphics->EnableRendering();
```
So basically this is going to prevent the JS thread from doing anything during a call to `EnableRendering`? I think this won't be necessary after @bghgary's in flight changes that remove auto rendering and makes graphics threading much more robust. We'll probably need to wait for that change to wrap up and then see how this part of the change would apply (if at all).

nkari82 replied:
Blocks JavaScript threads of Babylon js mapped to bgfx.

I don't actually use it. I wasted time safely shutting down and recreating bgfx. Even if I shut it down safely, it crashes by bgfx related call mapped with babylonjs before bgfx is created again (so it doesn't shut down bgfx).

chrisfromwork replied:
thanks for the context @nkari82!

ryantrem commented on Modules/@babylonjs/react-native/shared/BabylonNative.h:
```diff
@@ -8,8 +8,10 @@ namespace Babylon
-    void UpdateView(void* windowPtr, size_t width, size_t height, void* windowTypePtr = nullptr);
+    void UpdateView(void* windowPtr, size_t width, size_t height, void* windowTypePtr = nullptr, void* contextPtr = nullptr, void* backBufferPtr = nullptr);
     void RenderView();
+    void EnableView();
+    void DisableView();
```
Are these actually used? Maybe I just missed it, not sure!

nkari82 replied:
This code is not used.

chrisfromwork commented:
I've been spending some more time looking at this review. If we don't end up listening/acting on composition scale changes for the swap chain panel well consistently draw content with the correct size. However, it seems like something about how this is all set up prevents anti aliasing from working correctly:

![Screenshot (86)](https://user-images.githubusercontent.com/10225640/109722426-877b9700-7b61-11eb-9afa-34c2e01eaf48.png)
![Screenshot (87)](https://user-images.githubusercontent.com/10225640/109722434-89455a80-7b61-11eb-8a73-55b3202da8e2.png)

TBD why anti aliasing doesn't seem to be working. I'll provide more information here as we determine whats going on. @nkari82  if you have any initial thoughts on why this may be happening definitely let us know. :)

rgerd commented:
If you want to have antialiasing, you could do it similarly to how bgfx (BabylonNative's rendering backend) does it:

[They create a texture outside of the swapchain that supports MSAA](https://github.com/bkaradzic/bgfx/blob/master/src/renderer_d3d11.cpp#L1083-L1101)

[They use that MSAA texture in their render target view](https://github.com/bkaradzic/bgfx/blob/master/src/renderer_d3d11.cpp#L2221)

([Which would be analogous to this line in your PR](https://github.com/BabylonJS/BabylonReactNative/pull/160/files#diff-59eecabc40d052e377bb76ad16e57d580211d2c27a11b60f471c7687cb10ea29R431))

[You'll want to be sure you create a multisampled depth/stencil buffer as well, because I'm pretty sure this code won't run with multisampling in the sample description when creating the depth/stencil buffer for you](https://github.com/bkaradzic/bgfx/blob/master/src/renderer_d3d11.cpp#L2231-L2249)

[And then when they're finished they resolve the MSAA render target into their swapchain color buffer](https://github.com/bkaradzic/bgfx/blob/master/src/renderer_d3d11.cpp#L6578-L6579), which I assume you would do after the call to `RenderView()`.

--

[Here's a description of MSAA for those who haven't been exposed to it yet](https://learnopengl.com/Advanced-OpenGL/Anti-Aliasing)

[And a parallel description on how to do multisampling in a generic UWP app](https://docs.microsoft.com/en-us/windows/uwp/gaming/multisampling--multi-sample-anti-aliasing--in-windows-store-apps)

--

None of this is tested, just an educated guess. I could spend some time proofing it out since it should be pretty straightforward to attempt.

nkari82 commented:
here! msaa
![msaa](https://user-images.githubusercontent.com/15668659/109784941-ed921800-7c4e-11eb-9816-d5bb5cb23582.png)

rgerd commented on Modules/@babylonjs/react-native-windows/windows/BabylonReactNative/EngineView.cpp:
```diff
@@ -163,12 +237,279 @@ namespace winrt::BabylonReactNative::implementation {
+
+        if (_sampleDesc.Count > 1)
+        {
+            ::Microsoft::WRL::ComPtr<ID3D11Texture2D> originBackBuffer, msaaTexutePtr;
```
`msaaTexturePtr`

rgerd commented on Modules/@babylonjs/react-native-windows/windows/BabylonReactNative/EngineView.h:
```diff
@@ -1,37 +1,59 @@
+
+        DXGI_SAMPLE_DESC _sampleDesc{ 1, 0 };
+        ::Microsoft::WRL::ComPtr<IDXGISwapChain2>_swapChain;
+        ::Microsoft::WRL::ComPtr <ID3D11RenderTargetView> _backBufferPtr;
```
```suggestion
        ::Microsoft::WRL::ComPtr<ID3D11RenderTargetView> _backBufferPtr;
```

(Same on below line)

rgerd commented on Modules/@babylonjs/react-native/shared/BabylonNative.cpp:
```diff
@@ -142,7 +144,7 @@ namespace Babylon
             });
         }
-
+       			
```
(nit, remove whitespace to avoid use of tabs as well as file diff churn)
```suggestion

```

rgerd commented on Modules/@babylonjs/react-native/shared/BabylonNative.cpp:
```diff
@@ -278,7 +280,7 @@ namespace Babylon
         }
     }
-
+	
```
(nit, remove whitespace to avoid use of tabs as well as file diff churn)
```suggestion

```

rgerd commented on Modules/@babylonjs/react-native-windows/windows/BabylonReactNative/EngineView.cpp:
```diff
@@ -163,12 +237,279 @@ namespace winrt::BabylonReactNative::implementation {
+        }
+        else // Otherwise, create a new one
+        {
+            _sampleDesc = s_msaa[static_cast<uint32_t>(MSAA::X4)];  // MSAA::X4 Default Setting in BabylonNative
```
Seems like this is fine for now, but it might be worth filing an issue to eventually make this externally configurable, since I could see BNRN developers potentially wanting control over the MSAA level to balance performance / video memory / visual quality.

nkari82 replied:
that's right

chrisfromwork commented:
@rgerd and @nkari82 although this logic works, we may want to assess whether we can introduce the fixes into bgfx. This change in its current form seems like it is now reimplementing a good amount of logic already in bgfx to support MSAA.

@nkari82 it seems like a few of the underlying crashes are related to the SwapChainPanel still holding a reference to the underlying swap chain when its released by bgfx (this happens during both resize operations and engine tear down). Are there any other clear issues with bgfx in regards to hot reload that this review is addressing?

nkari82 commented:
@chrisfromwork
There are two problems.
1. Crash in bgfx in swapchain when resizing
2. 3d object is not visible during hot reloading. (I modified and tested bgfx regarding the swap chain.)

I don't want to fix bgfx if there is a way to get around the problem.

The amount of code is still simple, so it's okay, but if the code grows more, I think a way to fix bgfx would be okay.

chrisfromwork commented:
FWIW, @nkari82 I am going to work to get this change checked in as well: https://github.com/BabylonJS/bgfx/pull/15

nkari82 commented:
@chrisfromwork

I looked at the changes in bgfx.

https://github.com/bkaradzic/bgfx/blob/master/src/dxgi.cpp#L592

What do you think about "swapChainPanelNative->SetSwapChain(NULL)" behavior at this location when a destructor is called from EngineView? Then I don't think I need a "removeSwapChain" :)

chrisfromwork commented:
> @chrisfromwork
> 
> I looked at the changes in bgfx.
> 
> https://github.com/bkaradzic/bgfx/blob/master/src/dxgi.cpp#L592
> 
> What do you think about "swapChainPanelNative->SetSwapChain(NULL)" behavior at this location when a destructor and OnSizeChanged is called from EngineView? Then I don't think I need a "removeSwapChain" :)

we can definitely do this

> @chrisfromwork
> 
> I looked at the changes in bgfx.
> 
> https://github.com/bkaradzic/bgfx/blob/master/src/dxgi.cpp#L592
> 
> What do you think about "swapChainPanelNative->SetSwapChain(NULL)" behavior at this location when a destructor and OnSizeChanged is called from EngineView? Then I don't think I need a "removeSwapChain" :)

After looking at this on my end there were a few bugs I saw:
* bgfx did not remove the swap chain from the swap chain panel when destroying it (this is now fixed)
* bgfx did not update the SwapChainPanel reference in its renderer after initialization. This made it so destroying/creating new EngineViews did not work correctly (this is now fixed)
* the react-native-windows Abi safe jsi runtime is currently destroyed prior to cleaning up the underlying jsi objects/jsi runtime. This results in javascript logic continuing to execute when things are in a bad state, which leads to crashes. This will be fixed with this pr: https://github.com/microsoft/react-native-windows/pull/7356

We could do this SetSwapChain call on our end, but it probably isn't necessary now that bgfx has fixes. That all going said, these different fixes I have introduced avoid crashes but have resulted in white panel flashing when the swap chain panel is changed. There is still room for improvement here specifically related to continuing to show babylon.js content during resize/fast refresh scenarios

nkari82 commented:
> > @chrisfromwork
> > I looked at the changes in bgfx.
> > https://github.com/bkaradzic/bgfx/blob/master/src/dxgi.cpp#L592
> > What do you think about "swapChainPanelNative->SetSwapChain(NULL)" behavior at this location when a destructor and OnSizeChanged is called from EngineView? Then I don't think I need a "removeSwapChain" :)
> 
> we can definitely do this
> 
> > @chrisfromwork
> > I looked at the changes in bgfx.
> > https://github.com/bkaradzic/bgfx/blob/master/src/dxgi.cpp#L592
> > What do you think about "swapChainPanelNative->SetSwapChain(NULL)" behavior at this location when a destructor and OnSizeChanged is called from EngineView? Then I don't think I need a "removeSwapChain" :)
> 
> After looking at this on my end there were a few bugs I saw:
> 
> * bgfx did not remove the swap chain from the swap chain panel when destroying it (this is now fixed)
> * bgfx did not update the SwapChainPanel reference in its renderer after initialization. This made it so destroying/creating new EngineViews did not work correctly (this is now fixed)
> * the react-native-windows Abi safe jsi runtime is currently destroyed prior to cleaning up the underlying jsi objects/jsi runtime. This results in javascript logic continuing to execute when things are in a bad state, which leads to crashes. This will be fixed with this pr: [microsoft/react-native-windows#7356](https://github.com/microsoft/react-native-windows/pull/7356)
> 
> We could do this SetSwapChain call on our end, but it probably isn't necessary now that bgfx has fixes. That all going said, these different fixes I have introduced avoid crashes but have resulted in white panel flashing when the swap chain panel is changed. There is still room for improvement here specifically related to continuing to show babylon.js content during resize/fast refresh scenarios

thanks for sharering~

chrisfromwork commented:
@nkari82 at this time we think we have a working fast refresh experience for react-native-windows. There are still some race conditions around loading textures in BabylonNative when tearing down the jsi runtime, but we will be working through those as well. Are there other items that you feel are still needed for react-native-windows fast refresh?

nkari82 commented:
> @nkari82 at this time we think we have a working fast refresh experience for react-native-windows. There are still some race conditions around loading textures in BabylonNative when tearing down the jsi runtime, but we will be working through those as well. Are there other items that you feel are still needed for react-native-windows fast refresh?

chrisfromwork commented:
@nkari82 that's super cool that you have a typescript bgfx wrapper working! 

I haven't spent a ton of time looking into custom asset loaders/shader compilers/texture converters. But we definitely could add these to the issues list if you want to file feature requests. :)
